crawler.c

```diff
@@ -278,28 +285,81 @@ static void crawler_metadump_eval(crawler_module_t *cm, item *it, uint32_t hv, i
     }
     client_flags_t flags;
     FLAGS_CONV(it, flags);
-    // TODO: uriencode directly into the buffer.
-    uriencode(ITEM_key(it), keybuf, it->nkey, KEY_MAX_URI_ENCODED_LENGTH);
-    int total = snprintf(cm->c.buf + cm->c.bufused, 4096,
-            "key=%s exp=%ld la=%llu cas=%llu fetch=%s cls=%u size=%lu flags=%llu\n",
-            keybuf,
-            (it->exptime == 0) ? -1 : (long)(it->exptime + process_started),
-            (unsigned long long)(it->time + process_started),
-            (unsigned long long)ITEM_get_cas(it),
-            (it->it_flags & ITEM_FETCHED) ? "yes" : "no",
-            ITEM_clsid(it),
-            (unsigned long) ITEM_ntotal(it),
-            (unsigned long long)flags);
-    refcount_decr(it);
-    // TODO: some way of tracking the errors. these should be impossible given
-    // the space requirements.
-    if (total >= LRU_CRAWLER_MINBUFSPACE - 1 || total <= 0) {
-        // Failed to write, don't push it.
-        return;
+    assert(it->nkey * 3 < LRU_CRAWLER_MINBUFSPACE*2);
+    // unrolled snprintf for ~30% time improvement on fullspeed dump
+    // key=%s exp=%ld la=%llu cas=%llu fetch=%s cls=%u size=%lu flags=%llu\n
+    // + optional ext_page=%u and ext_offset=%u
+
+    char *p = cm->c.buf + cm->c.bufused;
+    char *start = p;
+    KADD(p, "key=", 4);
```

Comment:
nit: wouldn't it be better to have static strings instead of magic strings.

Reply:
would be inconsistent with rest of code + it ends up being a lot more verbose. not really, no.